ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java

```diff
@@ -3,51 +3,118 @@
+  private final int botPositionYIndex = 2;
+  private final int botRotationIndex = 5;
+  private final int noTagInSightId = -1;
+  NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight");
+  /**
+   * a constructor for the camera subsystem class
+   * @param subsystems - the subsystem collection
+   */
+  public CameraSubsystem() {
+  }
 
   /**
-   * Example command factory method.
-   *
-   * @return a command
+   * a method that returns a vision measurement. 
+   * pose portion of the vision measurement is null if there is no valid measurement. 
    */
-  public Command exampleMethodCommand() {
-    // Inline construction of command goes here.
-    // Subsystem::RunOnce implicitly requires `this` subsystem.
-    return runOnce(
-        () -> {
-          /* one-time action goes here */
-        });
+  public VisionMeasurement getVisionBotPose(){
+    double tagId = table.getEntry("tid").getDouble(0);
+
+    if (tagId == noTagInSightId){
+      return new VisionMeasurement(null, 0.0);
+    }
+    else{
+      double[] botpose = table.getEntry("botpose").getDoubleArray(new double[defaultDoubleArraySize]);
+      Double timestamp = Timer.getFPGATimestamp() - (botpose[TimestampIndex]/milisecondsInSeconds);
+      Translation2d botTranslation = new Translation2d(botpose[botPositionXIndex], botpose[botPositionYIndex]);
+      Rotation2d botYaw = Rotation2d.fromDegrees(botpose[botRotationIndex]);
+      Pose2d realRobotPosition = new Pose2d(botTranslation, botYaw);
+      return new VisionMeasurement(realRobotPosition, timestamp);
+    }
+  }
+
+  public double getTagId(){
+    return table.getEntry("tid").getDouble(0);
+  }
+
+  public DistanceMeasurement getDistanceFromTag(double blueTId, double redTId){
+    DistanceMeasurement measurement = new DistanceMeasurement(false, 0.0);
+    if((getTagId() == blueTId || getTagId() == redTId) && getVisionBotPoseInTargetSpace() != null){
+      measurement.setIsValid(true);
+
+      double xDistance = getVisionBotPoseInTargetSpace().getTranslation().getX();
+      double yDistance = getVisionBotPoseInTargetSpace().getTranslation().getY();
+      double totalDistance = Math.sqrt(Math.pow(xDistance, 2) + Math.pow(yDistance, 2));
+      measurement.setDistanceMeteres(totalDistance);
+    }
+    return measurement;
   }
 
   /**
-   * An example method querying a boolean state of the subsystem (for example, a digital sensor).
-   *
-   * @return value of some boolean subsystem state, such as a digital sensor.
+   * a method that returns a vision measurement. 
+   * pose portion of the vision measurement is null if there is no valid measurement. 
    */
-  public boolean exampleCondition() {
-    // Query some boolean state, such as a digital sensor.
-    return false;
+  public Pose2d getVisionBotPoseInTargetSpace(){
+    double tagId = table.getEntry("tid").getDouble(0);
+    double[] botpose = table.getEntry("botpose_targetspace").getDoubleArray(new double[defaultDoubleArraySize]);
```

Comment:
While seemingly not required, using 'this.' keyword when referencing member variables helps the reader understand that the variable here is not stack, but instead its heap (member).

Reply:
as in ...

double[] botpose = **this.**table.getEntry("botpose_targetspace").getDoubleArray(new double[**this.**defaultDoubleArraySize]);

Reply:
added this's everywhere